core/src/main/scala/cats/Traverse.scala

```diff
@@ -284,4 +287,35 @@ object Traverse {
   @deprecated("Use cats.syntax object imports", "2.2.0")
   object nonInheritedOps extends ToTraverseOps
 
+  private[cats] def traverseDirectly[Coll[x] <: IterableOnce[x], G[_], A, B](
+    builder: mutable.Builder[B, Coll[B]]
+  )(fa: IterableOnce[A])(f: A => G[B])(implicit G: StackSafeMonad[G]): G[Coll[B]] = {
+    val size = fa.knownSize
+    if (size >= 0) {
+      builder.sizeHint(size)
+    }
+    G.map(fa.iterator.foldLeft(G.pure(builder)) { case (accG, a) =>
+      G.flatMap(accG) { acc =>
+        G.map(f(a)) { a =>
+          acc += a
+          acc
+        }
+      }
+    })(_.result())
+  }
+
+  private[cats] def traverse_Directly[G[_], A, B](
+    fa: IterableOnce[A]
+  )(f: A => G[B])(implicit G: StackSafeMonad[G]): G[Unit] = {
+    val iter = fa.iterator
+    if (iter.hasNext) {
+      val first = iter.next()
+      G.map(iter.foldLeft(f(first)) { case (g, a) =>
```

Comment:
I think this can be G.void(.. so that has a chance for an optimized implementation.

Reply:
Yes, good catch!